File: backend/app/routes/events.py
from datetime import datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Query
from fastapi.responses import Response

from app.core.deps import CurrentUserId, DbSession
from app.repositories.events import EventWithMeta
from app.schemas.events import (
    AttendeeListResponse,
    AttendeeRead,
    EventCreate,
    EventListResponse,
    EventRead,
    RsvpRead,
)
from app.services.events import EventService

router = APIRouter(prefix="/events", tags=["events"])


def _to_read(item: EventWithMeta) -> EventRead:
    return EventRead(
        id=item.event.id,
        organizer_id=item.event.organizer_id,
        organizer_name=item.organizer_name,
        title=item.event.title,
        description=item.event.description,
        category=item.event.category,
        location=item.event.location,
        starts_at=item.event.starts_at,
        ends_at=item.event.ends_at,
        max_attendees=item.event.max_attendees,
        attendee_count=item.attendee_count,
        created_at=item.event.created_at,
        updated_at=item.event.updated_at,
    )


@router.get("", response_model=EventListResponse)
def list_events(
    db: DbSession,
    _user_id: CurrentUserId,
    category: Annotated[str | None, Query()] = None,
    from_: Annotated[datetime | None, Query(alias="from")] = None,
    to: Annotated[datetime | None, Query()] = None,
    q: Annotated[str | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> EventListResponse:
    result = EventService(db).list(
        category=category,
        from_dt=from_,
        to_dt=to,
        q=q,
        limit=limit,
        offset=offset,
    )
    return EventListResponse(
        items=[_to_read(item) for item in result.items],
        total=result.total,
    )


@router.post("", response_model=EventRead, status_code=201)
def create_event(
    payload: EventCreate,
    db: DbSession,
    user_id: CurrentUserId,
) -> EventRead:
    item = EventService(db).create(
        organizer_id=user_id,
        title=payload.title,
        description=payload.description,
        category=payload.category,
        location=payload.location,
        starts_at=payload.starts_at,
        ends_at=payload.ends_at,
        max_attendees=payload.max_attendees,
    )
    return _to_read(item)


@router.get("/{event_id}", response_model=EventRead)
def get_event(
    event_id: UUID, db: DbSession, _user_id: CurrentUserId
) -> EventRead:
    return _to_read(EventService(db).get(event_id))


@router.delete("/{event_id}", status_code=204)
def delete_event(
    event_id: UUID, db: DbSession, user_id: CurrentUserId
) -> Response:
    EventService(db).delete(event_id=event_id, requester_id=user_id)
    return Response(status_code=204)


@router.post(
    "/{event_id}/attendees", response_model=RsvpRead, status_code=201
)
def rsvp(
    event_id: UUID, db: DbSession, user_id: CurrentUserId
) -> RsvpRead:
    attendance = EventService(db).rsvp(event_id=event_id, user_id=user_id)
    return RsvpRead(
        event_id=attendance.event_id,
        user_id=attendance.user_id,
        created_at=attendance.created_at,
    )


@router.delete("/{event_id}/attendees/me", status_code=204)
def cancel_rsvp(
    event_id: UUID, db: DbSession, user_id: CurrentUserId
) -> Response:
    EventService(db).cancel_rsvp(event_id=event_id, user_id=user_id)
    return Response(status_code=204)


@router.get("/{event_id}/attendees", response_model=AttendeeListResponse)
def list_attendees(
    event_id: UUID, db: DbSession, _user_id: CurrentUserId
) -> AttendeeListResponse:
    rows = EventService(db).list_attendees(event_id=event_id)
    return AttendeeListResponse(
        items=[
            AttendeeRead(
                user_id=r.user_id,
                display_name=r.display_name,
                rsvp_at=r.rsvp_at,
            )
            for r in rows
        ],
        total=len(rows),
    )

File: frontend/src/lib/api/events.ts
import { apiFetch } from "@/lib/api";
import { getToken } from "@/lib/auth";

export type EventCategory = "academic" | "social" | "sports" | "culture" | "other";

export type EventRead = {
  id: string;
  organizer_id: string;
  organizer_name: string;
  organizer_avatar_url: string | null;
  title: string;
  description: string | null;
  category: EventCategory;
  location: string | null;
  starts_at: string;
  ends_at: string | null;
  max_attendees: number | null;
  attendee_count: number;
  created_at: string;
  updated_at: string;
};

export type EventListResponse = {
  items: EventRead[];
  total: number;
};

export type EventFilters = {
  category?: string;
  from?: string;
  to?: string;
  q?: string;
  mine?: boolean;
  limit?: number;
  offset?: number;
};

export type EventCreateInput = {
  title: string;
  description?: string | null;
  category: EventCategory;
  location?: string | null;
  starts_at: string;
  ends_at?: string | null;
  max_attendees?: number | null;
};

export type Attendee = {
  user_id: string;
  display_name: string;
  avatar_url: string | null;
  rsvp_at: string;
};

export type AttendeeListResponse = {
  items: Attendee[];
  total: number;
};

export type RsvpRead = {
  event_id: string;
  user_id: string;
  created_at: string;
};

function buildQuery(f: EventFilters): string {
  const params = new URLSearchParams();
  if (f.category) params.set("category", f.category);
  if (f.from) params.set("from", f.from);
  if (f.to) params.set("to", f.to);
  if (f.q) params.set("q", f.q);
  if (f.mine) params.set("mine", "1");
  if (f.limit !== undefined) params.set("limit", String(f.limit));
  if (f.offset !== undefined) params.set("offset", String(f.offset));
  const query = params.toString();
  return query ? `?${query}` : "";
}

export async function listEvents(filters: EventFilters = {}): Promise<EventListResponse> {
  return apiFetch<EventListResponse>(`/events${buildQuery(filters)}`, {
    method: "GET",
    token: getToken(),
  });
}

export async function getEvent(id: string): Promise<EventRead> {
  return apiFetch<EventRead>(`/events/${id}`, {
    method: "GET",
    token: getToken(),
  });
}

export async function createEvent(input: EventCreateInput): Promise<EventRead> {
  return apiFetch<EventRead>("/events", {
    method: "POST",
    body: input,
    token: getToken(),
  });
}

export async function deleteEvent(id: string): Promise<void> {
  await apiFetch<null>(`/events/${id}`, {
    method: "DELETE",
    token: getToken(),
  });
}

export async function rsvpEvent(id: string): Promise<RsvpRead> {
  return apiFetch<RsvpRead>(`/events/${id}/attendees`, {
    method: "POST",
    token: getToken(),
  });
}

export async function cancelRsvp(id: string): Promise<void> {
  await apiFetch<null>(`/events/${id}/attendees/me`, {
    method: "DELETE",
    token: getToken(),
  });
}

export async function listAttendees(id: string): Promise<AttendeeListResponse> {
  return apiFetch<AttendeeListResponse>(`/events/${id}/attendees`, {
    method: "GET",
    token: getToken(),
  });
}
